README.md
msgpack-cli
===========

msgpack-cli is command line tool that converts data from JSON to [Msgpack](http://msgpack.org) and vice versa. Also allows calling RPC methods via [msgpack-rpc](https://github.com/msgpack-rpc/msgpack-rpc/blob/master/spec.md).

Installation
------------

    go get github.com/jakm/msgpack-cli

Usage
-----

    msgpack-cli

    Usage:
        msgpack-cli encode <input-file> [--out=<output-file>] [--disable-int64-conv]
        msgpack-cli decode <input-file> [--out=<output-file>] [--pp]
        msgpack-cli rpc <host> <port> <method> [<params>|--file=<input-file>] [--pp]
            [--timeout=<timeout>][--disable-int64-conv]
        msgpack-cli -h | --help
        msgpack-cli --version

    Commands:
        encode                Encode data from input file to STDOUT
        decode                Decode data from input file to STDOUT
        rpc                   Call RPC method and write result to STDOUT

    Options:
        -h --help             Show this help message and exit
        --version             Show version
        --out=<output-file>   Write output data to file instead of STDOUT
        --file=<input-file>   File where parameters or RPC method are read from
        --pp                  Pretty-print - indent output JSON data
        --timeout=<timeout>   Timeout of RPC call [default: 30]
        --disable-int64-conv  Disable the default behaviour such that JSON numbers
                              are converted to float64 or int64 numbers by their
                              meaning, all result numbers will have float64 type


    Arguments:
        <input-file>          File where data are read from
        <host>                Server hostname
        <port>                Server port
        <method>              Name of RPC method
        <params>              Parameters of RPC method in JSON format

Examples
--------

Encoding/decoding:

    $ cat test.json
    {
      "firstName": "John",
      "lastName": "Smith",
      "isAlive": true,
      "age": 25,
      "height_cm": 167.6,
      "address": {
        "streetAddress": "21 2nd Street",
        "city": "New York",
        "state": "NY",
        "postalCode": "10021-3100"
      },
      "phoneNumbers": [
        {
          "type": "home",
          "number": "212 555-1234"
        },
        {
          "type": "office",
          "number": "646 555-4567"
        }
      ],
      "children": [],
      "spouse": null
    }
    $
    $ msgpack-cli encode test.json --out test.bin
    $
    $ ls -l test.* | awk '{print $9, $5}'
    test.bin 242
    test.json 429
    $
    $ msgpack-cli decode test.bin --pp  # pretty-print
    {
      "address": {
        "city": "New York",
        "postalCode": "10021-3100",
        "state": "NY",
        "streetAddress": "21 2nd Street"
      },
      "age": 25,
      "children": [],
      "firstName": "John",
      "height_cm": 167.6,
      "isAlive": true,
      "lastName": "Smith",
      "phoneNumbers": [
        {
          "number": "212 555-1234",
          "type": "home"
        },
        {
          "number": "646 555-4567",
          "type": "office"
        }
      ],
      "spouse": null
    }

RPC calling:

    $ # zero params
    $ msgpack-cli rpc localhost 8000 echo
    []
    $
    $ # single param
    $ msgpack-cli rpc localhost 8000 echo 3.159
    [3.159]
    $
    $ # multiple params (as json array)
    $ msgpack-cli rpc localhost 8000 echo '["abc", "def", "ghi", {"A": 65, "B": 66, "C": 67}]'
    ["abc","def","ghi",{"A":65,"B":66,"C":67}]



debian/changelog
msgpack-cli (0.4) unstable; urgency=low

  * Fixed interoperability issue
  * Renamed main source file
  * Added test of convertNumberTypes
  * convertNumberTypes rewritten to use pointer to object

 -- Jakub Matys <jakub.matys@nodus.cz>  Thu, 12 Feb 2015 19:37:29 +0100

msgpack-cli (0.3) unstable; urgency=low

  * Improved build automation
  * The connection making moved outside callRPC
  * Rename: doEncDec -> doConversion
  * The rpc command can read parameters from file
  * Readme updated
  * Added --pp option
  * Added --timeout option
  * Readme updated

 -- Jakub Matys <jakub.matys@nodus.cz>  Thu, 16 Oct 2014 07:40:21 +0200

msgpack-cli (0.2) unstable; urgency=low

  * Debian packaging
  * Build script
  * Added test scripts

 -- Jakub Matys <jakub.matys@nodus.cz>  Wed, 08 Oct 2014 07:15:51 +0200

msgpack-cli (0.1) unstable; urgency=low

  * Initial release

 -- Jakub Matys <matys.jakub@gmail.com>  Tue, 07 Oct 2014 08:00:37 +0200


test/run.sh
#!/bin/bash

clean_test() {
    [ "$RPC_SERVER_PID" != "" ] && kill $RPC_SERVER_PID 2> /dev/null
    [ "$TESTDIR" != "" ] && rm -rf $TESTDIR
}

fail() {
    clean_test
    echo
    echo -e "\e[31mFAILED: ${1}\e[39m"
    exit 1
}

cancel() {
    fail "CTRL-C detected"
}

info() {
    echo -e "\e[34m${1}\e[39m"
    echo
}

trap cancel INT

# -------------------

info "Preparing test environment..."

TESTDIR=$(mktemp -d)

virtualenv $TESTDIR > /dev/null || fail "virtualenv creation"

cd $(dirname $0)/..
go build -o $TESTDIR/bin/msgpack-cli || fail "msgpack-cli build"

cp test/assert-json-equal $TESTDIR/bin
cp test/assert-msgpack-equal $TESTDIR/bin
cp test/rpc-server $TESTDIR/bin

cp test/data.json $TESTDIR
cp test/data.bin $TESTDIR

cd $TESTDIR

. $TESTDIR/bin/activate

pip install msgpack-rpc-python > /dev/null || fail "msgpack-rpc-python installation"

echo "Starting rpc-server on localhost:8000..."
echo

./bin/rpc-server &

RPC_SERVER_PID=$(echo $!)

sleep 1

if ! kill -0 $RPC_SERVER_PID; then
    fail "rpc-server is not running"
fi

# -------------------

info "Testing RPC..."

echo "RPC: echo"
assert-json-equal "$(msgpack-cli rpc localhost 8000 echo)" "[]"

echo "RPC: echo 3.14159"
assert-json-equal "$(msgpack-cli rpc localhost 8000 echo 3.14159)" "[3.14159]"

echo "RPC: echo test"
assert-json-equal "$(msgpack-cli rpc localhost 8000 echo text)" '["text"]'

echo "RPC: echo \"long test\""
assert-json-equal "$(msgpack-cli rpc localhost 8000 echo "long text")" '["long text"]'

echo "RPC: echo '[\"abc\", \"def\", \"ghi\", {\"A\": 65, \"B\": 66, \"C\": 67}]'"
assert-json-equal "$(msgpack-cli rpc localhost 8000 echo '["abc", "def", "ghi", {"A": 65, "B": 66, "C": 67}]')" '["abc","def","ghi",{"A":65,"B":66,"C":67}]'

# -------------------

info "Testing encoding/decoding..."

echo "Encode: data.json"
msgpack-cli encode data.json --out=output.bin
assert-msgpack-equal output.bin data.bin

echo "Decode: data.bin"
assert-json-equal "$(msgpack-cli decode data.bin)" "$(cat data.json)"

# -------------------

clean_test


main_test.go
// Copyright 2015 Jakub Matys
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
    "encoding/json"
    "reflect"
    "testing"
)

func TestConvertNumberTypes(t *testing.T) {
    testConversionOfScalarValue(t)
    testRecursiveConversionOfSlice(t)
    testRecursiveConversionOfMap(t)
}

func testConversionOfScalarValue(t *testing.T) {
    var (
        err    error
        object interface{}
    )

    for _, num := range []string{"0.0", ".0", "-4.", "3.14159", "-4.8e-8", "34e15"} {
        object = json.Number(num)

        err = convertNumberTypes(&object)

        if err != nil {
            t.Fatalf("Conversion of value \"%s\" failed: %s", num, err)
        }

        if reflect.TypeOf(object).Kind() != reflect.Float64 {
            t.Fatalf("Value \"%s\" was was converted to %s type (float64 expected).",
                num, reflect.TypeOf(object))
        }
    }

    for _, num := range []string{"0", "256", "-3", "123456789", "1234567890000"} {
        object = json.Number(num)

        err = convertNumberTypes(&object)

        if err != nil {
            t.Fatalf("Conversion of value \"%s\" failed: %s", num, err)
        }

        if reflect.TypeOf(object).Kind() != reflect.Int64 {
            t.Fatalf("Value \"%s\" was was converted to %s type (int64 expected).",
                num, reflect.TypeOf(object))
        }
    }
}

func testRecursiveConversionOfSlice(t *testing.T) {
    var object interface{}

    object = []interface{}{
        "string",
        json.Number("1234567890000"),
        []interface{}{
            json.Number("256"),
            json.Number(".0"),
            "string",
        },
    }

    checkType := func(object interface{}, expectedType reflect.Kind) {
        if reflect.TypeOf(object).Kind() != expectedType {
            t.Fatalf("Type %s of value %v doesn't match expected type %s",
                reflect.TypeOf(object), object, expectedType)
        }
    }

    if err := convertNumberTypes(&object); err != nil {
        t.Fatalf("Conversion of slice %v failed: %s", object, err)
    } else {
        switch object := object.(type) {
        case []interface{}:
            checkType(object[0], reflect.String)
            checkType(object[1], reflect.Int64)
            checkType(object[2], reflect.Slice)
            switch inner := object[2].(type) {
            case []interface{}:
                checkType(inner[0], reflect.Int64)
                checkType(inner[1], reflect.Float64)
                checkType(inner[2], reflect.String)
            default:
                t.Fatalf("Conversion function returned incorrect type %s of inner slice (expected: []interface{})",
                    reflect.TypeOf(inner))
            }
        default:
            t.Fatalf("Conversion function returned incorrect type %s (expected: []interface{})",
                reflect.TypeOf(object))
        }
    }
}

func testRecursiveConversionOfMap(t *testing.T) {
    var object interface{}

    object = map[string]interface{}{
        "aaa": "string",
        "bbb": json.Number("1234567890000"),
        "ccc": json.Number("4."),
        "ddd": map[string]interface{}{
            "xxx": json.Number("256"),
            "yyy": json.Number(".0"),
            "zzz": "string",
        },
    }

    checkType := func(object interface{}, expectedType reflect.Kind) {
        if reflect.TypeOf(object).Kind() != expectedType {
            t.Fatalf("Type %s of value %v doesn't match expected type %s",
                reflect.TypeOf(object), object, expectedType)
        }
    }

    if err := convertNumberTypes(&object); err != nil {
        t.Fatalf("Conversion of map %v failed: %s", object, err)
    } else {
        switch object := object.(type) {
        case map[string]interface{}:
            checkType(object["aaa"], reflect.String)
            checkType(object["bbb"], reflect.Int64)
            checkType(object["ccc"], reflect.Float64)
            checkType(object["ddd"], reflect.Map)
            switch inner := object["ddd"].(type) {
            case map[string]interface{}:
                checkType(inner["xxx"], reflect.Int64)
                checkType(inner["yyy"], reflect.Float64)
                checkType(inner["zzz"], reflect.String)
            default:
                t.Fatalf("Conversion function returned incorrect type %s of inner map (expected: map[string]interface{})",
                    reflect.TypeOf(inner))
            }
        default:
            t.Fatalf("Conversion function returned incorrect type %s (expected: map[string]interface{})",
                reflect.TypeOf(object))
        }
    }
}


msgpack-cli.go
// Copyright 2014 Jakub Matys
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
    "encoding/json"
    "fmt"
    "github.com/docopt/docopt-go"
    "github.com/ugorji/go/codec"
    "io/ioutil"
    "log"
    "net"
    "net/rpc"
    "os"
    "reflect"
    "strconv"
    "strings"
    "time"
    "unicode"
    "unicode/utf8"
)

const usage = `msgpack-cli

Usage:
    msgpack-cli encode <input-file> [--out=<output-file>]
    msgpack-cli decode <input-file> [--out=<output-file>] [--pp]
    msgpack-cli rpc <host> <port> <method> [<params>|--file=<input-file>] [--pp] [--timeout=<timeout>]
    msgpack-cli -h | --help
    msgpack-cli --version

Commands:
    encode               Encode data from input file to STDOUT
    decode               Decode data from input file to STDOUT
    rpc                  Call RPC method and write result to STDOUT

Options:
    -h --help            Show this help message and exit
    --version            Show version
    --out=<output-file>  Write output data to file instead of STDOUT
    --file=<input-file>  File where parameters or RPC method are read from
    --pp                 Pretty-print - indent output JSON data
    --timeout=<timeout>  Timeout of RPC call [default: 30]

Arguments:
    <input-file>         File where data are read from
    <host>               Server hostname
    <port>               Server port
    <method>             Name of RPC method
    <params>             Parameters of RPC method in JSON format`

type Options struct {
    indent  bool
    timeout uint32
}

type RPCResult struct {
    reply interface{}
    err   error
}

func main() {
    arguments, err := docopt.Parse(usage, nil, true, "msgpack-cli "+__VERSION__, false)
    if err != nil {
        log.Fatal(err)
    }

    switch {
    case arguments["encode"], arguments["decode"]:
        inFilename := arguments["<input-file>"].(string)
        outFilename, _ := arguments["--out"].(string)

        convertFunc := convertJSON2Msgpack
        if arguments["decode"].(bool) {
            convertFunc = convertMsgpack2JSON
        }

        options := Options{indent: arguments["--pp"].(bool)}

        err = doConversion(inFilename, outFilename, convertFunc, options)
    case arguments["rpc"]:
        host := arguments["<host>"].(string)
        port := arguments["<port>"].(string)
        method := arguments["<method>"].(string)
        var params string
        params, err = getRPCParams(arguments)
        if err != nil {
            break
        }
        var timeout uint32
        timeout, err = getTimeout(arguments)
        if err != nil {
            break
        }

        options := Options{
            indent:  arguments["--pp"].(bool),
            timeout: timeout,
        }

        err = doRPC(host, port, method, params, options)
    default:
        panic("unreachable")
    }
    if err != nil {
        log.Fatal(err)
    }
}

func doConversion(inFilename, outFilename string, convertFunc func(data []byte, options Options) ([]byte, error), options Options) error {
    inFile, err := os.Open(inFilename)
    if err != nil {
        return err
    }
    defer inFile.Close()

    outFile := os.Stdout
    if outFilename != "" {
        if outFile, err = os.Create(outFilename); err != nil {
            return err
        }
        defer outFile.Close()
    }

    var inBuffer, outBuffer []byte

    if inBuffer, err = ioutil.ReadAll(inFile); err != nil {
        return fmt.Errorf("Reading error: %s", err)
    }

    if outBuffer, err = convertFunc(inBuffer, options); err != nil {
        return err
    }

    var n int

    if n, err = outFile.Write(outBuffer); err != nil {
        return fmt.Errorf("Writing error: %s", err)
    }
    if n != len(outBuffer) {
        return fmt.Errorf("Writing error: written %d of %d bytes", n, len(outBuffer))
    }

    return nil
}

func doRPC(host, port, method, params string, options Options) error {
    if len(params) == 0 {
        params = "[]"
    } else if !strings.HasPrefix(params, "[") {
        if char, _ := utf8.DecodeRuneInString(params); unicode.IsLetter(char) {
            params = strconv.Quote(params)
        }
        params = "[" + params + "]"
    }

    args, err := decodeJSON(params)
    if err != nil {
        return err
    }

    var conn net.Conn
    if conn, err = net.Dial("tcp", host+":"+port); err != nil {
        return err
    }
    defer conn.Close()

    result := make(chan RPCResult)
    defer close(result)

    go callRPC(result, conn, method, args)

    select {
    case res := <-result:
        if res.err != nil {
            return res.err
        }

        var jsonData string
        if jsonData, err = encodeJSON(res.reply, options.indent); err != nil {
            return err
        }

        fmt.Println(jsonData)
    case <-time.After(time.Duration(options.timeout) * time.Second):
        return fmt.Errorf("RPC call timed out")
    }

    return nil
}

func callRPC(result chan<- RPCResult, conn net.Conn, method string, args interface{}) {
    handle := getHandle()
    rpcCodec := codec.MsgpackSpecRpc.ClientCodec(conn, &handle)
    client := rpc.NewClientWithCodec(rpcCodec)

    var reply interface{}
    var mArgs codec.MsgpackSpecRpcMultiArgs = args.([]interface{})

    if err := client.Call(method, mArgs, &reply); err != nil {
        result <- RPCResult{reply: nil, err: fmt.Errorf("RPC error: %s", err)}
    }

    result <- RPCResult{reply: reply, err: nil}
}

func convertJSON2Msgpack(data []byte, options Options) (result []byte, err error) {
    var object interface{}

    if object, err = decodeJSON(string(data)); err != nil {
        return nil, err
    }

    if result, err = encodeMsgpack(object); err != nil {
        return nil, err
    }

    return result, nil
}

func convertMsgpack2JSON(data []byte, options Options) (result []byte, err error) {
    var object interface{}

    if object, err = decodeMsgpack(data); err != nil {
        return nil, err
    }

    var jsonData string
    if jsonData, err = encodeJSON(object, options.indent); err != nil {
        return nil, err
    }
    result = []byte(jsonData)

    return result, nil
}

func getHandle() (handle codec.MsgpackHandle) {
    handle = codec.MsgpackHandle{RawToString: true}
    handle.MapType = reflect.TypeOf(map[string]interface{}(nil))
    return handle
}

func encodeJSON(object interface{}, indent bool) (data string, err error) {
    var bytes []byte
    if indent {
        bytes, err = json.MarshalIndent(object, "", "  ")
    } else {
        bytes, err = json.Marshal(object)
    }
    if err != nil {
        return "", fmt.Errorf("JSON encoding: %s", err)
    }
    return string(bytes), nil
}

func decodeJSON(data string) (object interface{}, err error) {
    bytes := []byte(data)
    if err := json.Unmarshal(bytes, &object); err != nil {
        return nil, fmt.Errorf("JSON decoding: %s", err)
    }
    return object, nil
}

func encodeMsgpack(object interface{}) (data []byte, err error) {
    handle := getHandle()
    encoder := codec.NewEncoderBytes(&data, &handle)
    if err := encoder.Encode(object); err != nil {
        return nil, fmt.Errorf("Msgpack encoding: %s", err)
    }
    return data, err
}

func decodeMsgpack(data []byte) (object interface{}, err error) {
    handle := getHandle()
    decoder := codec.NewDecoderBytes(data, &handle)
    if err := decoder.Decode(&object); err != nil {
        return nil, fmt.Errorf("Msgpack decoding: %s", err)
    }
    return object, err
}

func getRPCParams(arguments map[string]interface{}) (params string, err error) {
    params, _ = arguments["<params>"].(string)
    filename, _ := arguments["--file"].(string)

    if filename != "" {
        buff, err := ioutil.ReadFile(filename)
        if err != nil {
            return "", err
        }
        params = string(buff)
    }

    return params, nil
}

func getTimeout(arguments map[string]interface{}) (timeout uint32, err error) {
    timeout = uint32(30)
    if str := arguments["--timeout"].(string); str != "" {
        var tmp uint64
        tmp, err = strconv.ParseUint(str, 10, 32)
        timeout = uint32(tmp)
    }
    return timeout, err
}


main.go
// Copyright 2014 Jakub Matys
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
    "bytes"
    "encoding/json"
    "fmt"
    "github.com/docopt/docopt-go"
    "github.com/ugorji/go/codec"
    "io/ioutil"
    "log"
    "net"
    "net/rpc"
    "os"
    "reflect"
    "strconv"
    "strings"
    "time"
    "unicode"
    "unicode/utf8"
)

const usage = `msgpack-cli

Usage:
    msgpack-cli encode <input-file> [--out=<output-file>] [--disable-int64-conv]
    msgpack-cli decode <input-file> [--out=<output-file>] [--pp]
    msgpack-cli rpc <host> <port> <method> [<params>|--file=<input-file>] [--pp]
        [--timeout=<timeout>][--disable-int64-conv]
    msgpack-cli -h | --help
    msgpack-cli --version

Commands:
    encode                Encode data from input file to STDOUT
    decode                Decode data from input file to STDOUT
    rpc                   Call RPC method and write result to STDOUT

Options:
    -h --help             Show this help message and exit
    --version             Show version
    --out=<output-file>   Write output data to file instead of STDOUT
    --file=<input-file>   File where parameters or RPC method are read from
    --pp                  Pretty-print - indent output JSON data
    --timeout=<timeout>   Timeout of RPC call [default: 30]
    --disable-int64-conv  Disable the default behaviour such that JSON numbers
                          are converted to float64 or int64 numbers by their
                          meaning, all result numbers will have float64 type


Arguments:
    <input-file>          File where data are read from
    <host>                Server hostname
    <port>                Server port
    <method>              Name of RPC method
    <params>              Parameters of RPC method in JSON format`

type Options struct {
    convertToInt64 bool
    indent         bool
    timeout        uint32
}

type RPCResult struct {
    reply interface{}
    err   error
}

func main() {
    arguments, err := docopt.Parse(usage, nil, true, "msgpack-cli "+__VERSION__, false)
    if err != nil {
        log.Fatal(fmt.Errorf("Arguments parsing: %s", err))
    }

    switch {
    case arguments["encode"], arguments["decode"]:
        inFilename := arguments["<input-file>"].(string)
        outFilename, _ := arguments["--out"].(string)

        convertFunc := convertJSON2Msgpack
        if arguments["decode"].(bool) {
            convertFunc = convertMsgpack2JSON
        }

        options := Options{
            convertToInt64: !arguments["--disable-int64-conv"].(bool),
            indent:         arguments["--pp"].(bool),
        }

        err = doConversion(inFilename, outFilename, convertFunc, options)
    case arguments["rpc"]:
        host := arguments["<host>"].(string)
        port := arguments["<port>"].(string)
        method := arguments["<method>"].(string)
        var params string
        params, err = getRPCParams(arguments)
        if err != nil {
            break
        }
        var timeout uint32
        timeout, err = getTimeout(arguments)
        if err != nil {
            break
        }

        options := Options{
            convertToInt64: !arguments["--disable-int64-conv"].(bool),
            indent:         arguments["--pp"].(bool),
            timeout:        timeout,
        }

        err = doRPC(host, port, method, params, options)
    default:
        panic("unreachable")
    }
    if err != nil {
        log.Fatal(err)
    }
}

func doConversion(inFilename, outFilename string, convertFunc func(data []byte, options Options) ([]byte, error), options Options) error {
    inFile, err := os.Open(inFilename)
    if err != nil {
        return err
    }
    defer inFile.Close()

    outFile := os.Stdout
    if outFilename != "" {
        if outFile, err = os.Create(outFilename); err != nil {
            return err
        }
        defer outFile.Close()
    }

    var inBuffer, outBuffer []byte

    if inBuffer, err = ioutil.ReadAll(inFile); err != nil {
        return fmt.Errorf("Reading error: %s", err)
    }

    if outBuffer, err = convertFunc(inBuffer, options); err != nil {
        return err
    }

    var n int

    if n, err = outFile.Write(outBuffer); err != nil {
        return fmt.Errorf("Writing error: %s", err)
    }
    if n != len(outBuffer) {
        return fmt.Errorf("Writing error: written %d of %d bytes", n, len(outBuffer))
    }

    return nil
}

func doRPC(host, port, method, params string, options Options) error {
    if len(params) == 0 {
        params = "[]"
    } else if !strings.HasPrefix(params, "[") {
        if char, _ := utf8.DecodeRuneInString(params); unicode.IsLetter(char) {
            params = strconv.Quote(params)
        }
        params = "[" + params + "]"
    }

    args, err := decodeJSON(params, options.convertToInt64)
    if err != nil {
        return err
    }

    var conn net.Conn
    if conn, err = net.Dial("tcp", host+":"+port); err != nil {
        return err
    }
    defer conn.Close()

    result := make(chan RPCResult)
    defer close(result)

    go callRPC(result, conn, method, args)

    select {
    case res := <-result:
        if res.err != nil {
            return res.err
        }

        var jsonData string
        if jsonData, err = encodeJSON(res.reply, options.indent); err != nil {
            return err
        }

        fmt.Println(jsonData)
    case <-time.After(time.Duration(options.timeout) * time.Second):
        return fmt.Errorf("RPC call timed out")
    }

    return nil
}

func callRPC(result chan<- RPCResult, conn net.Conn, method string, args interface{}) {
    handle := getHandle()
    rpcCodec := codec.MsgpackSpecRpc.ClientCodec(conn, &handle)
    client := rpc.NewClientWithCodec(rpcCodec)

    var reply interface{}
    var mArgs codec.MsgpackSpecRpcMultiArgs = args.([]interface{})

    if err := client.Call(method, mArgs, &reply); err != nil {
        result <- RPCResult{reply: nil, err: fmt.Errorf("RPC error: %s", err)}
    }

    result <- RPCResult{reply: reply, err: nil}
}

func convertJSON2Msgpack(data []byte, options Options) (result []byte, err error) {
    var object interface{}

    if object, err = decodeJSON(string(data), options.convertToInt64); err != nil {
        return nil, err
    }

    if result, err = encodeMsgpack(object); err != nil {
        return nil, err
    }

    return result, nil
}

func convertMsgpack2JSON(data []byte, options Options) (result []byte, err error) {
    var object interface{}

    if object, err = decodeMsgpack(data); err != nil {
        return nil, err
    }

    var jsonData string
    if jsonData, err = encodeJSON(object, options.indent); err != nil {
        return nil, err
    }
    result = []byte(jsonData)

    return result, nil
}

func getHandle() (handle codec.MsgpackHandle) {
    handle = codec.MsgpackHandle{RawToString: true}
    handle.MapType = reflect.TypeOf(map[string]interface{}(nil))
    return handle
}

func encodeJSON(object interface{}, indent bool) (data string, err error) {
    var buffer bytes.Buffer
    encoder := json.NewEncoder(&buffer)
    if err := encoder.Encode(object); err != nil {
        return "", fmt.Errorf("JSON encoding: %s", err)
    }
    data = buffer.String()

    if indent {
        buffer.Truncate(0)
        if err = json.Indent(&buffer, []byte(data), "", "  "); err != nil {
            return "", fmt.Errorf("JSON encoding: %s", err)
        }
        data = buffer.String()
    }

    return data, nil
}

func decodeJSON(data string, convertToInt64 bool) (object interface{}, err error) {
    decoder := json.NewDecoder(strings.NewReader(data))

    if convertToInt64 {
        decoder.UseNumber()
    }

    if err = decoder.Decode(&object); err != nil {
        return nil, fmt.Errorf("JSON decoding: %s", err)
    }

    if convertToInt64 {
        if err = convertNumberTypes(&object); err != nil {
            return nil, fmt.Errorf("JSON decoding: %s", err)
        }
    }
    return object, nil
}

func encodeMsgpack(object interface{}) (data []byte, err error) {
    handle := getHandle()
    encoder := codec.NewEncoderBytes(&data, &handle)
    if err := encoder.Encode(object); err != nil {
        return nil, fmt.Errorf("Msgpack encoding: %s", err)
    }
    return data, err
}

func decodeMsgpack(data []byte) (object interface{}, err error) {
    handle := getHandle()
    decoder := codec.NewDecoderBytes(data, &handle)
    if err := decoder.Decode(&object); err != nil {
        return nil, fmt.Errorf("Msgpack decoding: %s", err)
    }
    return object, err
}

func convertNumberTypes(object *interface{}) (err error) {
    switch value := (*object).(type) {
    case json.Number:
        // fmt.Printf("Type %s, value %v\n", reflect.TypeOf(value), value)
        if strings.ContainsAny(value.String(), ".eE") {
            *object, err = value.Float64()
        } else {
            *object, err = value.Int64()
        }
    case []interface{}:
        // fmt.Printf("Type %s, value %v\n", reflect.TypeOf(value), value)
        for idx := range value {
            if err = convertNumberTypes(&value[idx]); err != nil {
                break
            }
        }
    case map[string]interface{}:
        // fmt.Printf("Type %s, value %v\n", reflect.TypeOf(value), value)
        for k, v := range value {
            if err = convertNumberTypes(&v); err != nil {
                break
            } else {
                value[k] = v
            }
        }
    default:
        // fmt.Printf("Type %s, value %v\n", reflect.TypeOf(value), value)
    }

    if err != nil {
        object = nil
    }

    return err
}

func getRPCParams(arguments map[string]interface{}) (params string, err error) {
    params, _ = arguments["<params>"].(string)
    filename, _ := arguments["--file"].(string)

    if filename != "" {
        buff, err := ioutil.ReadFile(filename)
        if err != nil {
            return "", err
        }
        params = string(buff)
    }

    return params, nil
}

func getTimeout(arguments map[string]interface{}) (timeout uint32, err error) {
    timeout = uint32(30)
    if str := arguments["--timeout"].(string); str != "" {
        var tmp uint64
        tmp, err = strconv.ParseUint(str, 10, 32)
        timeout = uint32(tmp)
    }
    return timeout, err
}
